### PR TITLE
Assignment2 Matthew Adams - Rachael DeBoer

### PR DESCRIPTION
**Pull request by MattA41**
@rkdeboer and @MattA41 worked on this

### CODE REVIEW

**Comment by Matthew-Powers** [comment]
I will review this

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 1]
```diff
@@ -26,22 +26,64 @@ public void Error_WithData_LogsMessage()
         var logger = new TestLogger();
 
         // Act
-        //logger.Error("Message {0}", 42);
+        logger.Error("message 42");
```
Comment (Matthew-Powers) [inline Logger.Tests/BaseLoggerMixinsTests.cs:29]: Why message 42? Is 42 a magic number?

File: Logger/LogFactory.cs [hunk 2]
```diff
@@ -1,10 +1,19 @@
...
+
+namespace Logger;
 
 public class LogFactory
 {
-    public BaseLogger CreateLogger(string className)
+    private string? _filePath;
+    public BaseLogger? CreateLogger(string className)
     {
+        return _filePath == null ? null : (BaseLogger)new FileLogger(className);
```
Comment (Matthew-Powers) [inline Logger/LogFactory.cs:12]: Should not be directly accessing a private method like this

File: Logger/LogFactory.cs [hunk 3]
```diff
@@ -1,10 +1,19 @@
...
 
-        return null;
+    }
+    public void ConfigureFileLogger(string? filePath)
+    {
+        _filePath = filePath;
```
Comment (Matthew-Powers) [inline Logger/LogFactory.cs:17]: See other comment in this file

File: Logger/FileLogger.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Text;
+
+namespace Logger;
+    public class FileLogger(string filepath) : BaseLogger
```
Comment (Matthew-Powers) [inline Logger/FileLogger.cs:8]:
```suggestion
    public class FileLogger(string filePath) : BaseLogger
```

File: Logger/FileLogger.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+            this._logLevel = logLevel;
+            this._message = message;
+            File.AppendText(LogToString(logLevel, message, ClassName));
+
+        }
+
+        public static string LogToString(LogLevel logLevel, string message, string className)
+        {
+            string logString = DateTime.Now.ToString("MMM dd yyyy, hh:mm:ss", new CultureInfo("en-US"));
+            logString = logString + className + logLevel + message;
```
Comment (Matthew-Powers) [inline Logger/FileLogger.cs:36]:
```suggestion
            logString += className + logLevel + message;
```

File: Logger.Tests/BaseLoggerMixinsTests.cs (hunk 1 above)
Comment (MattA41) [reply Logger.Tests/BaseLoggerMixinsTests.cs:29]: To be honest I have no idea. I was working with the ta to determine why the tests were not working and would not work with any other formatting. the number 42 is a reference to the answer to life being 42. I can't remember the name of the book

File: Logger/FileLogger.cs (hunk 4 above)
Comment (MattA41) [reply Logger/FileLogger.cs:8]: thanks for the comment i missed that formating

File: Logger/FileLogger.cs (hunk 5 above)
Comment (MattA41) [reply Logger/FileLogger.cs:36]: I completely forgot about +=

File: Logger/LogFactory.cs (hunk 2 above)
Comment (MattA41) [reply Logger/LogFactory.cs:12]: we used to have it public but treated warnings as errors was getting mad about it. How would you have done it

**Review by MattA41 — COMMENTED** [review]
I looked at the comments and suggestions

File: Logger/LogFactory.cs (hunk 3 above)
Comment (StrangeRanger) [reply Logger/LogFactory.cs:17]: @Matthew-Powers Ideally, even if you have to repeat yourself, you shouldn't just say "See other comment in this file". It's better to be explicit when you are reviewing someone's pull request.

**Comment by StrangeRanger** [comment]
@Matthew-Power Make sure that all requirements that have been marked with an '❌', have their appropriate code commented on.

Example:

> Use the nameof() operator when identifying the class name to the logger ❌

If you think that the above requirement was not complete, make sure to comment on the code where `nameof()` should have been used.

File: Logger.Tests/BaseLoggerMixinsTests.cs (hunk 1 above)
Comment (BenjaminMichaelis) [reply Logger.Tests/BaseLoggerMixinsTests.cs:29]: The hitchhikers guide to the galaxy :)

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 6]
```diff
@@ -26,22 +26,64 @@ public void Error_WithData_LogsMessage()
...
+        Assert.AreEqual(LogLevel.Information, logger.LoggedMessages[0].LogLevel);
+        Assert.AreEqual("message 42", logger.LoggedMessages[0].Message);
     }
 
+    [TestMethod]
+    public void Debug_WithData_LogsMessage()
+    {
+        var logger = new TestLogger();
+
+        logger.Debug("message 42");
```
Comment (BenjaminMichaelis) [inline Logger.Tests/BaseLoggerMixinsTests.cs:66]: There should be a test taking in arguments

File: Logger.Tests/FileLoggerTests.cs [hunk 7]
```diff
@@ -1,9 +1,20 @@
...
+
+using Newtonsoft.Json.Serialization;
 
 namespace Logger.Tests;
 
 [TestClass]
 public class FileLoggerTests
 {
+    [TestMethod]
+    public void FileLogger_EmptyFilePath()
```
Comment (BenjaminMichaelis) [inline Logger.Tests/FileLoggerTests.cs:13]: I expect there need to be more tests than this, ones that actually test logging happening

File: Logger.Tests/FileLoggerTests.cs [hunk 8]
```diff
@@ -1,9 +1,20 @@
...
+    {
+        FileLogger fileLogger = new("");
+        Assert.IsFalse(fileLogger.FilePath == null);
```
Comment (BenjaminMichaelis) [inline Logger.Tests/FileLoggerTests.cs:16]: Assert.IsNull can be used here

File: Logger.Tests/Logger.Tests.csproj [hunk 9]
```diff
@@ -1,9 +1,14 @@
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
 
 	<PropertyGroup>
-		<TargetFramework>net7.0</TargetFramework>
+    <TargetFramework>net8.0</TargetFramework>
 		<IsPackable>false</IsPackable>
-		<LangVersion>11.0</LangVersion>
+		<LangVersion>12</LangVersion>
```
Comment (BenjaminMichaelis) [inline Logger.Tests/Logger.Tests.csproj:6]:
```suggestion
    <LangVersion>12.0</LangVersion>
```
12.0 is the current lang version. Also making spacing consistent with rest of the file

File: Logger/BaseLoggerMixins.cs [hunk 10]
```diff
@@ -1,6 +1,28 @@
-﻿namespace Logger;
+﻿using System;
+
+namespace Logger;
 
 public static class BaseLoggerMixins
 {
+    public static void Error(this BaseLogger? logger, string message, params object[] arguments)
```
Comment (BenjaminMichaelis) [inline Logger/BaseLoggerMixins.cs:7]: Why is logger nullable? I expect to call this message, a logger is needed.

File: Logger/FileLogger.cs [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+        public override string ClassName
+        {
+            get { return _className1; }
+            set { _className1 = value; }
+        }
+
+        public string FilePath { get => _filePath; set => _filePath = value; }
+        public string Message { get => _message; set => _message = value; }
+        public LogLevel LogLevel1 { get => _logLevel; set => _logLevel = value; }
```
Comment (BenjaminMichaelis) [inline Logger/FileLogger.cs:23]: Why is this named with the `1` suffix?
```suggestion
        public LogLevel LogLevel { get => _logLevel; set => _logLevel = value; }
```

File: Logger/FileLogger.cs [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+        public override void Log(LogLevel logLevel, string message)
+        {
+            this._logLevel = logLevel;
+            this._message = message;
```
Comment (BenjaminMichaelis) [inline Logger/FileLogger.cs:28]: We should reference these through the properties, not the fields

File: Logger/FileLogger.cs [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+using System.IO;
+using System.Text;
+
+namespace Logger;
+    public class FileLogger(string filePath) : BaseLogger
+    {
+        private LogLevel _logLevel;
+        private string _message = "No message given";
+        private string _className1 = "No class name given";
+        private string _filePath = filePath;
```
Comment (BenjaminMichaelis) [inline Logger/FileLogger.cs:13]: These should be properties

File: Logger/LogFactory.cs [hunk 14]
```diff
@@ -1,10 +1,19 @@
...
+﻿using System;
+using System.IO;
+using System.Runtime.CompilerServices;
+
+namespace Logger;
 
 public class LogFactory
 {
-    public BaseLogger CreateLogger(string className)
+    private string? _filePath;
```
Comment (BenjaminMichaelis) [inline Logger/LogFactory.cs:9]: This should be a property

File: Logger/LogFactory.cs (hunk 2 above)
Comment (BenjaminMichaelis) [inline Logger/LogFactory.cs:12]: Why is this cast needed?
```suggestion
        return _filePath == null ? null : new FileLogger(className);
```

File: Logger/Logger.csproj [hunk 15]
```diff
@@ -1,8 +1,13 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFramework>netstandard2.0</TargetFramework>
-    <LangVersion>11.0</LangVersion>
+    <TargetFramework>net8.0</TargetFramework>
+    <LangVersion>12</LangVersion>
```
Comment (BenjaminMichaelis) [inline Logger/Logger.csproj:5]:
```suggestion
    <LangVersion>12.0</LangVersion>
```

**Pull request closed without merging**